staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler.go

```diff
@@ -0,0 +1,339 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package peerproxy
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/rand"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+
+	"k8s.io/api/apiserverinternal/v1alpha1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	schema "k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/proxy"
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/storageversion"
+	apiserverproxyutil "k8s.io/apiserver/pkg/util/proxy"
+	apiserverinternallister "k8s.io/client-go/listers/apiserverinternal/v1alpha1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/transport"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/controlplane/reconcilers"
+)
+
+const (
+	PeerProxiedHeader = "peer-proxied"
+)
+
+type peerProxyHandler struct {
+	name                string
+	svLister            apiserverinternallister.StorageVersionLister
+	svi                 cache.SharedIndexInformer
+	svm                 storageversion.Manager
+	proxyClientCertFile string
+	proxyClientKeyFile  string
+	peerCAFile          string
+	svMap               sync.Map
+	lock                sync.RWMutex
```

Comment:
why do we need a sync.Map and a lock?
You already can access services in the  serviceInformers

Reply:
I see, svMap stores services group version map, add comment explaining these things so people don't have to read the code to understand it

Reply:
added godoc